File: frontend/lib/features/books/models/book.dart
import '../../auth/models/user.dart';
import 'book_index.dart';

/// Livro, espelhando o BookResource da API.
class Book {
  Book({
    required this.id,
    required this.titulo,
    required this.numeroPaginas,
    this.usuarioPublicador,
    this.indices = const [],
  });

  final int id;
  final String titulo;
  final int numeroPaginas;
  final User? usuarioPublicador;
  final List<BookIndex> indices;

  factory Book.fromJson(Map<String, dynamic> json) => Book(
        id: json['id'] as int,
        titulo: json['titulo'] as String? ?? '',
        numeroPaginas: json['numero_paginas'] as int? ?? 0,
        usuarioPublicador: json['usuario_publicador'] is Map<String, dynamic>
            ? User.fromJson(json['usuario_publicador'] as Map<String, dynamic>)
            : null,
        indices: (json['indices'] as List<dynamic>? ?? [])
            .map((e) => BookIndex.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

File: backend/app/Models/Book.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Book extends Model
{
    /** @use HasFactory<\Database\Factories\BookFactory> */
    use HasFactory;

    protected $fillable = [
        'user_id',
        'titulo',
        'numero_paginas',
    ];

    protected $casts = [
        'numero_paginas' => 'integer',
    ];

    /**
     * Usuario publicador do livro.
     *
     * @return BelongsTo<User, $this>
     */
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Todos os indices do livro (qualquer nivel da arvore).
     *
     * @return HasMany<BookIndex, $this>
     */
    public function indices(): HasMany
    {
        return $this->hasMany(BookIndex::class);
    }

    /**
     * Apenas os indices raiz (sem pai), ordenados por posicao.
     *
     * @return HasMany<BookIndex, $this>
     */
    public function rootIndices(): HasMany
    {
        return $this->hasMany(BookIndex::class)
            ->whereNull('parent_id')
            ->orderBy('position');
    }
}
